### ISSUE TITLE
raix:push Newbie Manual

### ISSUE THREAD

**Issue by harryward** [raix/push#115]
I just graduated from Newbie status and have utilized Raix in a few applications (one of them currently in apple and google play stores).

It took me a little while to get familiar with the apple certificate generation process, etc but once I did that it was super fast.

If you need any assistance getting raix:push setup or have questions throw them below and i'd be happy to help!

### The MOST IMPORTANT raix:push setup test
Most of you newbies are going to screw up the certificate generation process and its going to drive you crazy. Raix includes this link in his IOS certificate generation tutorial and I highly recommend utilizing this <a href="http://www.raywenderlich.com/32960/apple-push-notification-services-in-ios-6-tutorial-part-1">great writeup by Ali</a>.

The final step to generating your key, has you test your .pem files against the sandbox server (or if you want to test production just remove sandbox from the command below:

##### development test command
<pre>openssl s_client -connect gateway.sandbox.push.apple.com:2195 -cert PushChatCert.pem -key PushChatKey.pem</pre>

##### production test command
<pre>openssl s_client -connect gateway.push.apple.com:2195 -cert PushChatCert.pem -key PushChatKey.pem</pre>

Running the <b>test command</b> will prompt you to enter the password you entered when you exported your .p12 file.

##### command response
If you created the certificate and .pem files, etc correctly you are going to get a response like this. If you get an error, it means you did something wrong and your best bet is to start over.

<pre>
---
New, TLSv1/SSLv3, Cipher is AES256-SHA
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
SSL-Session:
    Protocol  : TLSv1
    Cipher    : AES256-SHA
    Session-ID: 
    Session-ID-ctx: 
    Master-Key: XXXXXXXXXXXXXXXXXXXX
    Key-Arg   : None
    Start Time: 1442776397
    Timeout   : 300 (sec)
    Verify return code: 0 (ok)
---
</pre>

If you get this response, you are golden.

##### Make sure you have a config.push.json
<pre>
{
  "apn": {
    "passphrase": "WHATEVER_PASSWORD_YOU_SET_ON_YOUR_P12_KEY",
    "key": "ios/apn-production/PushChatKey.pem",
    "cert": "ios/apn-production/PushChatCert.pem"
  },
  "apn-dev": {
    "passphrase": "WHATEVER_PASSWORD_YOU_SET_ON_YOUR_P12_KEY",
    "key": "iosGp/apn-dev/PushChatKey.pem",
    "cert": "iosGp/apn-dev/PushChatCert.pem"
  },
  "gcm": {
    "apiKey": "GET_THIS_FROM_GOOGLE_DEVELOPERS_CONSOLE",
    "projectNumber": GET_THIS_FROM_GOOGLE_DEVELOPERS_CONSOLE
  },
  "production":false,
  "badge": true,
  "sound": true,
  "alert": true,
  "vibrate": true
}
</pre>

##### Create a /server/pushMethods.js file to test
<pre>

Push.debug = true;

Push.allow({
    send: function(userId, notification) {
        return true; // Allow all users to send
    }
});

Meteor.methods({
    serverNotification: function(text,title) {
        var badge = 1
        Push.send({
            from: 'push',
            title: title,
            text: text,
            badge: badge,
            sound: 'airhorn.caf',
            payload: {
                title: title,
                text:text,
                historyId: result
            },
            query: {
                // this will send to all users
            }
        });
    },
    userNotification: function(text,title,userId) {
        var badge = 1
        Push.send({
            from: 'push',
            title: title,
            text: text,
            badge: badge,
            sound: 'airhorn.caf',
            payload: {
                title: title,
                historyId: result
            },
            query: {
                userId: userId //this will send to a specific Meteor.user()._id
            }
        });
    },
    removeHistory: function() {
        NotificationHistory.remove({}, function(error) {
            if (!error) {
                console.log("All history removed");
            }
        });
    },
});
</pre>

##### DEPLOY to blah.meteor server or utilize MODULUS
Your impatient instinct is going to make you jump to the app you have running on localhost and call the method you just created. DO NOT DO THAT.

<b>raix:push testing does not work from your local host</b>

Once you deploy your app to either a meteor or modulus server by running this...
<pre>meteor deploy blah.meteor.com</pre>
<pre>sudo modulus deploy</pre>

##### RUN YOUR APP with --mobile-settings

<pre>
meteor run ios-device --mobile-server blah.meteor.com --production
</pre>

once that runs its going to open up xCode - - plug in your device to your computer and select it from the available devices. (your device must have a provisioning profile - which is set up in the apple developer center <a href="https://developer.apple.com/account">https://developer.apple.com/account</a>.

Once you have the app succesfully launched on your phone, <b>it will prompt you saying 'Blah would like to send you push notifications<b>. If you see this, its a very good sign that this will work. Make sure you click 'YES' and once the app is loaded and everything looks good. <b>CLOSE YOUR APP</b>

<i>You cannot receive a push notification when the app is open on your phone, so whenever your testing make sure your app is closed.</i>

##### Execute your first test!!

1. navigate to your meteor web app (from computer web browser) (whatever you set --mobile-server to earlier is the url you need go to)

2. Open up your browsers developer console and click 'console' and run this command

<pre>Meteor.call('serverNotification','test','test')</pre>

It'll take a few seconds and you should see a push notification on your phone!!!

<img src="https://s3.amazonaws.com/f.cl.ly/items/1h2r0L3S0L0n400k1S3M/Screen%20Shot%202015-09-20%20at%2012.32.12%20PM.png">

**Comment by raix** [comment]
:+1: We should prop. add this to the docs folder

**Comment by harryward** [comment]
Sounds good thanks for the add on the readme @raix

**Comment by antonioandrade** [comment]
Very cool, thanks @harryward. Perhaps you should mention the collection creation
```javascript
NotificationHistory = new Mongo.Collection("notification_history");
```

**Comment by Chidan** [comment]
:+1: a well written tutorial, thanks for the effort, it sure helped me.

**Comment by albertor24** [comment]
What are the differences between doing this in ios and doing it in android? 

When I tried to do it, I could send the notifications to my device, but many were not received.

**Comment by harryward** [comment]
@antonioandrade good catch, goal was to eliminate all NotificationHistory references as that is just a nice to have and not required for this quick tutorial. I just removed that reference so that collection is not required.

**Comment by occasl** [comment]
Great tutorial @harryward . Going through it now. In the first section where you use openssl, you'll likely get "Verify return code: 20 (unable to get local issuer certificate)". If so, you need to download the [Entrust root cert] (https://www.entrust.net/downloads/root_request.cfm) and run the command as follows to get a return code of 0 (see [here](http://stackoverflow.com/questions/23343910/verify-errornum-20-when-connecting-to-gateway-sandbox-push-apple-com) for details):

```openssl s_client -connect gateway.sandbox.push.apple.com:2195 -cert cert-ma-approvals.pem -key key-ma-approvals.pem -CAfile entrust_2048_ca.cer```

**Comment by harryward** [comment]
@raix looks like my newbie manual got eliminated from the README is it out dated?

You can link to this from the README and i'll maintain it.
http://stached.io/standalone/fBLmRhsAuPSxKBSgM

**Comment by raix** [comment]
@harryward sorry about that - I think it happened when I branched of devel - I've added it back in and updated the link

**Comment by harryward** [comment]
Yep thats what I figured! your the man @raix thanks!

**Comment by LucNaterop** [comment]
I followed this along. All my keys were tested and are ok. In this tutorial it says that raix:push will not work on a locally hosted server. For me, apn ONLY works with localhost. If I deploy to meteor.com or modulus, I get the following server error: 

`Error: Socket unusable after connection. Hint: You may be using a certificate for the wrong environment`

Any ideas on what could be the problem?

**Comment by raix** [comment]
@LucNaterop https://github.com/raix/push/issues/134

**Comment by scsirdx** [comment]
Can't send push when using two different apps. I have one server app with web interface, and one cordova app, which connected to server using DDP.
I added apn settings and keys to server, added send method to server, and push initialisation in cordova app.
I have warning when starting cordova app, but no messages when i use push send from server. And no errors.

Any ideas?

**Comment by scsirdx** [comment]
Fixed it by using phonegap plugin and manually calling raix:* server methods (Meteor.connection.call()).

**Comment by raix** [comment]
@scsirdx the push package should be installed in both apps + servers - but if you just want one server to send you could set the send interval to `0` disabling the send.

**Comment by krishamoud** [comment]
What is `result` in this context?

**Comment by timbrandin** [comment]
@harryward I believe you need to remove result in your example on Stached, its undefined!
